genesis-tools/lockup-contract/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+//! A smart contract that allows lockup of a smart contract.
+use borsh::{BorshDeserialize, BorshSerialize};
+use key_management::{KeyType, PublicKey};
+use near_bindgen::{env, near_bindgen, Promise};
+use std::collections::{HashMap, HashSet};
+
+mod key_management;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct LockupContract {
+    lockup_amount: u128,
+    lockup_timestamp: u64,
+    keys: HashMap<KeyType, HashSet<PublicKey>>,
+    /// Whether this account is disallowed to stake anymore.
+    permanently_unstaked: bool,
```

Comment:
why would an account not be allowed to stake anymore?

Reply:
Please see the comment: https://github.com/nearprotocol/nearcore/blob/e76609702177767f2ca5d1e2b5a45006ac61b6e8/genesis-tools/lockup-contract/src/lib.rs#L209